File: lib/jsc/shell/JSString.cc
#include "flutter/lib/jsc/shell/JSString.h"
#include "JavaScriptCore/JSContextRef.h"
#include "JavaScriptCore/JSStringRef.h"

#include "third_party/dart/runtime/include/dart_api.h"
#include "third_party/tonic/converter/dart_converter.h"
#include "third_party/tonic/dart_args.h"
#include "third_party/tonic/dart_binding_macros.h"
#include "third_party/tonic/dart_library_natives.h"
#include "third_party/tonic/dart_persistent_value.h"
// #include "third_party/tonic/dart_microtask_queue.h"
#include "third_party/tonic/logging/dart_invoke.h"
// #include "third_party/tonic/typed_data/dart_byte_data.h"
#include "flutter/lib/jsc/shell/jsc_utils.h"

using tonic::ToDart;

namespace blink {

// RefCountedDartWrappable
// IMPLEMENT_WRAPPERTYPEINFO(jsc, JSValue);

long JSString::createWithCharacters(const std::u16string& str) {
    JSStringRef string = JSStringRetain(JSStringCreateWithCharacters(reinterpret_cast<const JSChar *>(str.data()), str.length()));
    return (long)string;
}

long JSString::createWithUTF8CString(std::string& str) {
    JSStringRef string = JSStringRetain(JSStringCreateWithUTF8CString(reinterpret_cast<const char*>(str.data())));
    return (long)string;
}

long JSString::retain(long strRef) {
    return (long) JSStringRetain((JSStringRef)strRef);
}

void JSString::release(long strRef) {
    JSStringRelease((JSStringRef)strRef);
}

long JSString::getLength(long strRef) {
  //TODO SIZE_T to long
    return (long)(JSStringGetLength((JSStringRef)strRef));
}

const char * JSString::toString(long strRef) {
    return reinterpret_cast<const char *>(JSStringGetCharactersPtr((JSStringRef)strRef));
    // char *buffer = new char[JSStringGetMaximumUTF8CStringSize((JSStringRef)stringRef)+1];
    // JSStringGetUTF8CString((JSStringRef)stringRef, buffer,
    //     JSStringGetMaximumUTF8CStringSize((JSStringRef)stringRef)+1);
    // jstring ret = env->NewStringUTF(buffer);
    // delete buffer;
    // return ret;
}

long JSString::getMaximumUTF8CStringSize(long strRef) {
    return (long) JSStringGetMaximumUTF8CStringSize((JSStringRef)strRef);
}

bool JSString::isEqual(long strRef,long strRef1 ) {
    return (bool) JSStringIsEqual((JSStringRef)strRef, (JSStringRef)strRef1);
}

bool JSString::isEqualToUTF8CString(long strRef,std::string& str) {
    return (bool) JSStringIsEqualToUTF8CString((JSStringRef)strRef, str.data());
}


#define FOR_EACH_BINDING(V) \
  V(JSString, createWithCharacters)       \
  V(JSString, createWithUTF8CString)   \
  V(JSString, retain)        \
  V(JSString, release)     \
  V(JSString, getLength)      \
  V(JSString, toString)      \
  V(JSString, getMaximumUTF8CStringSize)      \
  V(JSString, isEqual)      \
  V(JSString, isEqualToUTF8CString)
 
DART_BIND_ALL_STATIC(JSString, FOR_EACH_BINDING)

}  // namespace blink

File: lib/jsc/JSString.dart
part of dart.jsc;

class JSString extends JSValue {
  
  JSString([JSContext context, String value]) : super(context) {

  }

  @override
  // TODO:
  String toString() {
    return "";
  }

  @override
  // TODO:
  bool toBoolean() {
  }

  @override
  bool isNumber() {
    return false;
  }

  @override
  bool isString() {
    return true;
  }

  @override
  bool isBoolean() {
    // TODO: implement isBoolean
    return null;
  }

  @override
  bool isNull() {
    // TODO: implement isNull
    return null;
  }

  @override
  bool isObject() {
    // TODO: implement isObject
    return null;
  }

  @override
  bool isUndefined() {
    // TODO: implement isUndefined
    return null;
  }

  @override
  double toNumber() {
    // TODO: implement toNumber
    return null;
  }

  /// Gets the JSString value of this JS value
  /// @return  The JSString value
  JSString toJSString() {
    print('toJSString Entry $valueRef');
    CReturnValue cReturnValue = _toStringCopy(context.ctxRef(), valueRef);
    if (cReturnValue.exception != 0) {
      context.throwJSException(JSException.fromJSValue(
          JSValue.fromValueRef(cReturnValue.exception, context)));
      return null;
    }
    print('toJSString'+ cReturnValue.reference.toString());
    print('toJSString JSString'+ JSString(cReturnValue.reference).toString());
    return JSString(cReturnValue.reference);
  }


  int makeString(JSContext ctx, String value) native 'JSValue_makeString';

  //native function
  int createWithCharacters(String str) native 'JSString_createWithCharacters';

  int retain(int strRef) native 'JSString_retain';

  void release(int stringRef) native 'JSString_release';

  bool _isEqual(int a, int b) native 'JSString_isEqual';

  String _toString(int strRef) native 'JSString_toString';

  int getLength(int stringRef) native 'JSString_getLength';

  int getMaximumUTF8CStringSize(int stringRef) native 'JSString_getMaximumUTF8CStringSize';

  int createWithUTF8CString(String str) native 'JSString_createWithUTF8CString';

  bool isEqualToUTF8CString(int a, String b) native 'JSString_isEqualToUTF8CString';

}
